remote_attestation/rust/src/attestation.rs

```diff
@@ -15,238 +15,545 @@
+    /// Used for signing protocol transcripts and preventing replay attacks.
+    signer: Option<Signer>,
+}
+
+impl AttestationBehavior {
+    /// Represents an attestation process, where current machine remotely attests a remote peer and
+    /// verifies its attestation info.
+    pub fn create_peer_attestation(expected_tee_measurement: &[u8]) -> anyhow::Result<Self> {
         Ok(Self {
-            key_negotiator,
-            attestation_type: UnattestedSelf::new(tee_certificate),
+            expected_tee_measurement: Some(expected_tee_measurement.to_vec()),
+            tee_certificate: None,
+            signer: None,
         })
     }
-}
 
-impl AttestationEngine<UnattestedPeer> {
-    pub fn create(expected_tee_measurement: &[u8]) -> anyhow::Result<Self> {
-        let key_negotiator = KeyNegotiator::create().context("Couldn't create key negotiator")?;
+    /// Represents an attestation process, where current machine remotely attests to a remote peer
+    /// and sends attestation info to it.
+    pub fn create_self_attestation(tee_certificate: &[u8]) -> anyhow::Result<Self> {
         Ok(Self {
-            key_negotiator,
-            attestation_type: UnattestedPeer::new(expected_tee_measurement),
+            expected_tee_measurement: None,
+            tee_certificate: Some(tee_certificate.to_vec()),
+            signer: Some(Signer::create().context("Couldn't create signer")?),
         })
     }
-}
 
-impl AttestationEngine<MutuallyUnattested> {
-    pub fn create(tee_certificate: &[u8], expected_tee_measurement: &[u8]) -> anyhow::Result<Self> {
-        let key_negotiator = KeyNegotiator::create().context("Couldn't create key negotiator")?;
+    /// Represents an attestation process, where current machine and a remote peer remotely attest
+    /// each other.
+    pub fn create_bidirectional_attestation(
+        tee_certificate: &[u8],
+        expected_tee_measurement: &[u8],
+    ) -> anyhow::Result<Self> {
         Ok(Self {
-            key_negotiator,
-            attestation_type: MutuallyUnattested::new(tee_certificate, expected_tee_measurement),
+            expected_tee_measurement: Some(expected_tee_measurement.to_vec()),
+            tee_certificate: Some(tee_certificate.to_vec()),
+            signer: Some(Signer::create().context("Couldn't create signer")?),
         })
     }
-}
 
-/// Represents an attestation process, where current machine remotely attests to a remote peer and
-/// sends attestation info to it.
-pub struct UnattestedSelf {
-    /// PEM encoded X.509 certificate that signs TEE firmware key.
-    tee_certificate: Vec<u8>,
-}
+    pub fn contains_peer_attestation(&self) -> bool {
+        self.expected_tee_measurement.is_some()
+    }
 
-impl UnattestedSelf {
-    pub fn new(tee_certificate: &[u8]) -> Self {
-        Self {
-            tee_certificate: tee_certificate.to_vec(),
-        }
+    pub fn contains_self_attestation(&self) -> bool {
```

Comment:
What does self attestation mean? I don't understand at what stage of the handshake we can say that something contains self attestation. From the code it seems that only in the case of having self attestation the identity is added to the transcript, but I don't understand why.